tests/pytorch/distributed/run_layer_with_overlap.py

```diff
@@ -223,9 +286,27 @@ def _train(opts):
         shell=True,
     )
 
-    if result.stdout == "0":  # Extra checks for non-MNNVL platforms
+    if result.stdout == "0" and opts.tp is None:  # Extra checks for non-MNNVL platforms
         assert WORLD_SIZE == LOCAL_SIZE
 
+    # Initialize torch.distributed tp process group
+    new_group_kwargs = {
+        "backend": "nccl",
+    }
+    if opts.tp is not None:
+        LOCAL_SIZE = opts.tp
+        tp_base_rank = (WORLD_RANK // LOCAL_SIZE) * LOCAL_SIZE
+        tp_rank_list = list(range(tp_base_rank, tp_base_rank + LOCAL_SIZE))
+        new_group_kwargs = {
+            "backend": "nccl",
+            "ranks": tp_rank_list,
+        }
+    else:
+        opts.tp = WORLD_SIZE
+
+    assert opts.in_features is None or opts.layer_type is te.Linear
+    assert opts.out_features is None or opts.layer_type is te.LayerNormLinear
```

Comment:
I don't really like how strange and unintuitive the requirements are for `--in-features`/`--out-features`. I see we do it this way because we initialize UB based on `hidden_size`, but we should at least add comments and error messages to make it less mysterious. If we keep this API, we should also enforce its requirements for the parallel mode.
```suggestion
    # Tensor dim overrides for tensors that do not require TP communication
    if ops.in_features is not None:
        assert opts.layer_type is te.Linear and opts.linear_parallel_mode == "row, "--in-features is only used to configure row-tensor-parallel Linear layers. Use --num-heads or --head-dim for other cases." 
    if ops.out_features is not None:
        assert opts.layer_type is te.LayerNormLinear and opts.linear_parallel_mode == "column, "--in-features is only used to configure column-tensor-parallel LayerNormLinear layers. Use --num-heads or --head-dim for other cases."
```